arangod/Aql/OptimizerRules.cpp

```diff
@@ -1266,18 +1270,75 @@ void arangodb::aql::removeCollectVariablesRule(Optimizer* opt,
             auto usedThere =
                 Ast::getReferencedAttributesForKeep(exp->node(), outVariable,
                                                     isSafeForOptimization);
+
+
             if (isSafeForOptimization) {
               for (auto const& it : usedThere) {
                 keepAttributes.emplace(it);
               }
             } else {
               stop = true;
+              break;
+            }
+
+            if(other != nullptr) {
+
+              auto otherOutvar = other->outVariable();
+
+              // check if the name of the current variable turns up as attribute access in the next collect
+              bool isSafeForOptimization;
+              Ast::getReferencedAttributesForKeep(exp->node(), outVariable, isSafeForOptimization, true /*test indexed*/);
+              if(!isSafeForOptimization) {
+                stop = true;
+              }
+
+              // check if within the next collect the new group variable is used without any attribute
+              // if so the current variabe has to be available.
+              auto usedThere =
+                Ast::getReferencedAttributesForKeep(exp->node(), otherOutvar,
+                                                    isSafeForOptimization);
+              if(usedThere.empty()) {
```

Comment:
Why do we stop the optimization if there are no attributes used?!

Reply:
If no attributes are used we need the full variable.

Reply:
Yes. And if only a sub-attribute is used we don't seem to be detecting it and stop the optimization.